### PR TITLE
Refactored data clumps with the help of LLMs (research project)

### PR DESCRIPTION
**Pull request by compf**
Hello maintainers,

I am conducting a master thesis project focused on enhancing code quality through automated refactoring of data clumps, assisted by Large Language Models (LLMs).

<details>
  <summary>Data clump definition</summary>
  
A data clump exists if
1. two methods (in the same or in different classes) have at least 3 common parameters and one of those methods does not override the other,  or 
2. At least three fields in a class are common with the parameters of a method (in the same or in a different class), or
3. Two different classes have at least three common fields
  
See also the following UML diagram as an example
Example data clump
</details>

I believe these refactoring can contribute to the project by reducing complexity and enhancing readability of your source code.

Pursuant to the EU AI Act, I fully disclose the use of LLMs in generating these refactorings, emphasizing that all changes have undergone human review for quality assurance. 

Even if you decide not to integrate my changes to your codebase (which is perfectly fine), I ask you to fill out a feedback survey, which will be scientifically evaluated to determine the acceptance of AI-supported refactorings. You can find the feedback survey under https://campus.lamapoll.de/Data-clump-refactoring/en

Thank you for considering my contribution. I look forward to your feedback. If you have any other questions or comments, feel free to write a comment, or email me under tschoemaker@uni-osnabrueck.de .

Best regards,
Timo Schoemaker
Department of Computer Science
University of Osnabrück

-->

### Mandatory checks

- [ /] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ /] Tests created for changes (if applicable)
- [ /] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java [hunk 1]
```diff
@@ -52,16 +52,14 @@ public static Optional<XTextRange> getAnchor(XTextDocument doc, String name)
      * <p>
      * In LibreOffice the another name is in "{name}{number}" format.
      *
-     * @param name   For the bookmark.
-     * @param range  Cursor marking the location or range for the bookmark.
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:56]: The comments should moved to the new class

File: src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package org.jabref.model.openoffice.uno;
+
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+
+public class DocumentAnnotation{
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java:6]: This can be a record, cant it?

File: src/main/java/org/jabref/model/openoffice/uno/UnoNamed.java [hunk 3]
```diff
@@ -17,20 +17,14 @@ private UnoNamed() {
      * @param service For example "com.sun.star.text.ReferenceMark", "com.sun.star.text.Bookmark" or "com.sun.star.text.TextSection".
      *                <p>
      *                Passed to this.asXMultiServiceFactory().createInstance(service) The result is expected to support the XNamed and XTextContent interfaces.
-     * @param name    For the ReferenceMark, Bookmark, TextSection. If the name is already in use, LibreOffice may change the name.
-     * @param range   Marks the location or range for the thing to be inserted.
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/UnoNamed.java:21]: Check whether these comments can be moved to the new class, too.

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you for choosing JabRef. Some comments attached. 

IMHO LLMs cannot replace humans. It helps to have both Java and domain knowledge.

**Comment by koppor** [comment]
@compf please cite JabRef in your thesis. Use the citation as provided at https://docs.jabref.org/faq

**Comment by koppor** [comment]
@compf We have guideline to setup IntelliJ so that checkstyle won't complain at https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html. - Sorry, that this is such an effort. We did not dare to check in IntelliJ configs, because we fear that for each update of IntelliJ, the configs will change. We do not want to force our dozens of student contributors to fiddle around with their IntelliJ.

**Comment by compf** [comment]
Thank you very much for the feedback. I will certainly cite Jabref.

Somehow my Checkstyle was buggy so it didn't spot these formatting issues. After reinstalling, it worked

File: src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java [hunk 4]
```diff
@@ -101,8 +102,8 @@ private static void createReprInDocument(XTextDocument doc,
                 : left + right;
 
         cursor.getText().insertString(cursor, bracketedContent, true);
-
-        UnoReferenceMark.create(doc, refMarkName, cursor, true /* absorb */);
+        DocumentAnnotation docAnn = new DocumentAnnotation(doc, refMarkName, cursor, true /* absorb */);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java:105]: For a good refactoring `true` and `false` would be replaced by an `enum` containing `ABSORB` and (probably) `KEEP`. This is out of scope of this project, just as hint.

File: src/main/java/org/jabref/logic/openoffice/frontend/UpdateBibliography.java [hunk 5]
```diff
@@ -67,7 +68,8 @@ private static void createBibTextSection2(XTextDocument doc)
         // Alternatively, we could receive a cursor.
         XTextCursor textCursor = doc.getText().createTextCursor();
         textCursor.gotoEnd(false);
-        UnoTextSection.create(doc, BIB_SECTION_NAME, textCursor, false);
+        DocumentAnnotation annotation = new DocumentAnnotation(doc, BIB_SECTION_NAME, textCursor, false);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/frontend/UpdateBibliography.java:71]: Ah, LOL, the LLM puts different variable name for same variable types (`docAnn` versus `annotation` versus `documentAnnotation` below). I would propose to use always the full class name ( `documentAnnotation`) consistently.

A good LLM could recognize which variable name style is used and apply that style

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java [hunk 6]
```diff
@@ -52,16 +52,14 @@ public static Optional<XTextRange> getAnchor(XTextDocument doc, String name)
...
-     * @param absorb Shall we incorporate range?
      * @return The XNamed interface of the bookmark.
      * result.getName() should be checked by the caller, because its name may differ from the one requested.
+     * @param documentAnnotation The DocumentAnnotation object
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:57]: Can be removed, because the variable name also says that.

(Moreover, `@param` should have been put above `@return`. Our checkstyle seems to miss that configuration :p)

File: src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package org.jabref.model.openoffice.uno;
+
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+
+/**
+ * Represents a document annotation.
+ *
+ * @param doc    The document
+ * @param name   For the bookmark.
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java:10]: Please move the comment 

    For the ReferenceMark, Bookmark, TextSection. If the name is already in use, LibreOffice may change the name.

to here.

File: src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package org.jabref.model.openoffice.uno;
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java:1]: Move the class one level up. It just consists of classes residing in package `com.sun.star.text` and therefore is not directly part of `uno`, is it?

If it is, add a small comment to the package to explain why the class resides here.

(Your survery asked about that and therefore I am thinking more about it than I should :p)

File: src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+
+/**
+ * Represents a document annotation.
+ *
+ * @param doc    The document
+ * @param name   For the bookmark.
+ * @param range  Cursor marking the location or range for the bookmark.
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java:11]: Some comments say: `If an XTextSection by that name already exists, LibreOffice (6.4.6.2) creates a section with a name different from what we requested, in "Section {number}" format.`. Maybe, that should be put here, too? Or to the method getting this new object as parameter? - In all cases, the comment of the parameter should not get lost.

**Review by koppor — CHANGES_REQUESTED** [review]
Some more comments :)

I think, I should re-do the survey. The LLM is very bad in keeping `@param` comments. I, formyself, understood it now:

- There are `@param` comments for interpretation of the given object at each method
- They explain how the method behaves
- Thus, the comment should be moved into a newly created `@param` annotation for the called method.
- If the `@param` describes the parameter (and not the treatment of the parameter inside the method), it should be moved to the newly created class

(I hope, this was somehow clear)

File: src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java (hunk 8 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/model/openoffice/uno/DocumentAnnotation.java:1]: https://devdocs.jabref.org/code-howtos/openoffice/code-reorganization.html
uno : helpers for various tasks via UNO.
These are conceptually independent of JabRef code and logic.

**Comment by compf** [comment]
> Some more comments :)
> 
> I think, I should re-do the survey. The LLM is very bad in keeping `@param` comments. I, formyself, understood it now:
> 
> * There are `@param` comments for interpretation of the given object at each method
> * They explain how the method behaves
> * Thus, the comment should be moved into a newly created `@param` annotation for the called method.
> * If the `@param` describes the parameter (and not the treatment of the parameter inside the method), it should be moved to the newly created class
> 
> (I hope, this was somehow clear)

I think your first survey was fine. I have to admit that not everything was done by an LLM, in your case many things were performed by a tool I wrote, and which has still some strange bugs I need to fix, so I had to do a little manual refactoring (with associated mistakes), so you should not blame the LLM. 
I can try to fix the other issues if I find time, but keep in mind that I do not have an insight in your project like you do, and the feedback to the first commit is what's important for my study. Nevertheless, your projects is very interesting and I use Jabref often for managing  scientific sources, so maybe I will contribute in other ways :)

**Comment by koppor** [comment]
@compf Happy start into a new week. May I ask if you'll find time to finish this PR? 😅

**Review by koppor — APPROVED** [review]
